qiskit/transpiler/passes/utils/remove_final_measurements.py

```diff
@@ -19,12 +19,50 @@
 class RemoveFinalMeasurements(TransformationPass):
     """Remove final measurements and barriers at the end of a circuit.
 
-    This pass removes final barriers and final measurements, as well as the
-    ClassicalRegisters they are connected to if the ClassicalRegister
-    is unused. Measurements and barriers are considered final if they are
+    This pass removes final barriers and final measurements, as well as all
+    unused classical registers and bits they are connected to.
+    Measurements and barriers are considered final if they are
     followed by no other operations (aside from other measurements or barriers.)
+
+    Classical registers are removed iff they reference at least one bit
+    that has become unused by the circuit as a result of the operation, and all
+    of their other bits are also unused. Seperately, classical bits are removed
+    iff they have become unused by the circuit as a result of the operation,
+    or they appear in a removed classical register, but do not appear
+    in a classical register that will remain.
```

Comment:
That is diabolical. I believe what _should_ happen is the implicitly created 5-bit register is removed, along with `cr1`, because both of these registers are now idle as a result of the measure removal. However, `cr2` stays because it wasn't made idle due to the removal, it was idle beforehand. Therefore, we end up with `cr2` and its bits, only.

With the `c_if`, we'd coincidentally have the same result.

I'll write a test to confirm that behavior. Let me know if you wish it worked differently.

Reply:
Confirmed in 19943e94220a84b48a6a02b0f1b50a426e1b11fa.